### PR TITLE
When login-in via token, let a chance for user to set the password

### PR DESCRIPTION
**Pull request by Carreau**
When token is enabled, the login page will present a form to the user
asking them if they want to set a password at the same time. This is
almost equivalent to running `jupyter notebook password` on the command
line.

The experience can likely be better, but just submitting that as a POC
for feedback

### CODE REVIEW

File: notebook/auth/login.py [hunk 1]
```diff
@@ -72,16 +73,26 @@ def passwd_check(self, a, b):
...
+
+
+        
         if self.get_login_available(self.settings):
-            if self.passwd_check(self.hashed_password, typed_password):
+            if self.passwd_check(self.hashed_password, typed_password) and not new_password:
                 self.set_login_cookie(self, uuid.uuid4().hex)
             elif self.token and self.token == typed_password:
                 self.set_login_cookie(self, uuid.uuid4().hex)
+                if self.new_password:
```
Comment (minrk) [inline notebook/auth/login.py]: extra `self.`

File: notebook/templates/login.html [hunk 2]
```diff
@@ -85,6 +85,22 @@
       <p>
         Cookies are required for authenticated access to notebooks.
       </p>
+      <h3>{% trans %}Setup a Password{% endtrans %}</h3>
```
Comment (minrk) [inline notebook/templates/login.html]: This whole block should be conditional on password not being set already. And there should probably be a `set_password_available` for explicit disabling, as well.

**Comment by Carreau** [comment]
Yep I thought to most of those. I'm even considering redirecting users to
the login screen immediately if they set a password to be sure it works.

I'll work on that this week-end or next week. Thanks !

**Comment by Carreau** [comment]
Updated w/ documentation and options to disable.

File: notebook/notebookapp.py [hunk 3]
```diff
@@ -750,6 +751,18 @@ def _token_changed(self, change):
                       """
     )
 
+    allow_password_change = Bool(True, config=True, 
+                    help="""Allow password to be changed at login for the notebook server. 
+
+                    While login-in with a token, the notebook server UI will give the opportunity to
```
Comment (takluyver) [inline notebook/notebookapp.py]: logging in

File: notebook/templates/login.html [hunk 4]
```diff
@@ -85,6 +85,24 @@
...
+        <p> You can also setup a password by entering your token and a new password
+        on the fields below:</p>
+        <form action="{{base_url}}login?next={{next}}" method="post" class="">
+                {{ xsrf_form_html() | safe }}
+          <div class="form-group">
+            <input type="password" name="password" id="password_input" class="form-control" placeholder="Token">
+          </div>
+          <div class="form-group">
+            <input type="password" name="new_password" id="new_password_input"
+            class="form-control" placeholder="New password" required>
```
Comment (takluyver) [inline notebook/templates/login.html]: Should we get the user to type the new password twice, and check that it's the same?

**Comment by takluyver** [comment]
Can you show a screenshot of the new login page?

File: notebook/templates/login.html (hunk 4 above)
Comment (Carreau) [reply notebook/templates/login.html]: I don't think it is necessary, they can still issue `jupyter notebook password` to reset.

**Comment by Carreau** [comment]
In the end I think we want a change-password page that may or may not be available, and potentially pre-fill the token if it is given in the URL (and not redirect). 

Though I do not want to spend too much time on that as most user will not even see this change password field.

screen shot 2017-11-13 at 07 55 02

**Comment by takluyver** [comment]
Thanks. I'm happy to merge this and see how it goes, but I'll give it a while for other people to have a look.

File: docs/source/public_server.rst [hunk 5]
```diff
@@ -63,15 +63,28 @@ using the following command::
...
+
+You will be presented with a form asking for the current _token_, as well as
+your _new_ _password_ ; enter both and click on ``Login and setup new password``.
+
+Next time you need to log in you'll be able to use the new password instead of
+the login token, otherwise follow the procedure to set a password from the
+command line.
+
+The ability to change the password at first login time may be disabled by
+integrations by setting the ``--NotebookApp.allow_password_change=True``
```
Comment (minrk) [inline docs/source/public_server.rst]: =False

**Comment by Carreau** [comment]
> Minor comment that I think the flag for disabling password change is True where it should be False, but 👍 to merge with that typo fixed (or my understanding corrected).

Oops. Should be fixed.

**Pull request merged**